File: platforms/windows/src/TranslateTheDamn.Core/Backends/TranslatorRegistry.cs
using TranslateTheDamn.Core.Backends.Manifest;
using TranslateTheDamn.Core.Config;
using TranslateTheDamn.Core.Util;

namespace TranslateTheDamn.Core.Backends;

/// <summary>
/// Builds the backend id → <see cref="ITranslator"/> map by pairing the user's config with the
/// embedded declarative manifest (Constitution Q2): each backend is a generic
/// <see cref="ManifestCliBackend"/> / <see cref="ManifestHttpBackend"/> reading its definition from
/// <c>spec/backends.json</c>. Adding a backend = add a manifest entry (+ a config entry); no new code.
/// </summary>
public sealed class TranslatorRegistry
{
    private readonly Dictionary<string, ITranslator> _map = new(StringComparer.OrdinalIgnoreCase);

    public static TranslatorRegistry Build(AppConfig cfg)
    {
        var reg = new TranslatorRegistry();
        var manifest = BackendManifest.Load();
        // Resolve the unified target language once: {target} -> translation.targetLanguage. Every
        // prompt-driven backend (CLI + openai-http/anthropic-http) then shares the same target.
        var tmpl = PromptBuilder.WithTarget(cfg.Translation.PromptTemplate, cfg.Translation.TargetLanguage);

        foreach (var (id, bc) in cfg.Backends)
        {
            // Built-in backends resolve by id. A CUSTOM provider (id absent from the manifest) resolves a
            // generic HTTP template by its declared protocol — so user-typed base_url+key providers work
            // with no per-vendor manifest entry and no switch(id) (Constitution Law 6).
            if (!manifest.Backends.TryGetValue(id, out var def))
            {
                var tmplId = bc.Protocol?.Trim().ToLowerInvariant() switch
                {
                    "anthropic" => "anthropic-http",
                    "openai" => "openai-http",
                    _ => null
                };
                if (tmplId is null || !manifest.Backends.TryGetValue(tmplId, out def)) continue;
            }
            ITranslator? t = def.Kind.ToLowerInvariant() switch
            {
                "http" => new ManifestHttpBackend(id, def, bc, tmpl),
                "cli" => new ManifestCliBackend(id, def, bc, tmpl),
                _ => null
            };
            if (t is not null) reg._map[id] = t;
        }

        return reg;
    }

    /// <summary>Register or replace a translator (used for composition and tests).</summary>
    public void Add(ITranslator translator) => _map[translator.Id] = translator;

    public ITranslator? Get(string id) =>
        !string.IsNullOrEmpty(id) && _map.TryGetValue(id, out var t) ? t : null;

    public IReadOnlyCollection<string> Ids => _map.Keys;
}

File: platforms/macos/src/Core/TranslatorRegistry.swift
import Foundation

public final class TranslatorRegistry {
    private var map: [String: Translator] = [:]

    public init() {}

    public func register(_ translator: Translator, for id: String) {
        map[id.lowercased()] = translator
    }

    public func translator(for backend: String, config: BackendConfig, promptTemplate: String = "", runner: ProcessRunner = ProcessRunner()) -> Translator? {
        let key = backend.lowercased()
        if let existing = map[key] { return existing }

        // Built-in backends resolve by id. A CUSTOM provider (id absent from the manifest) resolves a
        // generic HTTP template by its declared protocol — so user-typed base_url+key providers work with
        // no per-vendor manifest entry and no switch(id) (Constitution Law 6). Mirrors Windows.
        var def = BackendManifest.backendDef(backend)
        var defId = backend
        if def == nil {
            let tmplId: String?
            switch (config.`protocol` ?? "").trimmingCharacters(in: .whitespaces).lowercased() {
            case "anthropic": tmplId = "anthropic-http"
            case "openai": tmplId = "openai-http"
            default: tmplId = nil
            }
            if let t = tmplId { def = BackendManifest.backendDef(t); defId = t }
        }
        guard let resolvedDef = def else { return nil }
        let kind = resolvedDef["kind"] as? String ?? ""

        let translator: Translator?

        switch kind.lowercased() {
        case "cli":
            translator = ProcessTranslator(id: backend, config: config, promptTemplate: promptTemplate, runner: runner)
        case "http":
            translator = HttpTranslator(id: backend, config: config, defId: defId, promptTemplate: promptTemplate)
        default:
            translator = nil
        }

        if let t = translator {
            map[key] = t
        }
        return translator
    }

    public var ids: [String] {
        return Array(map.keys)
    }
}
